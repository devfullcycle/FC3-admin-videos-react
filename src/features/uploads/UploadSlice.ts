import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { RootState } from "../../app/store";
import { updateVideo } from "./uploadThunk";

export const STATUS = {
  IDDLE: "idle",
  LOADING: "loading",
  SUCCESS: "success",
  FAILED: "failed",
} as const;

export type Status = typeof STATUS[keyof typeof STATUS];

export interface UploadState {
  id: string;
  file: File;
  field: string;
  videoId: string;
  progress?: number;
  status?: Status;
}

type UploadProgress = {
  id: string;
  progress: number;
};

const initialState: UploadState[] = [];

const uploadsSlice = createSlice({
  name: "uploads",
  initialState,
  reducers: {
    addUpload(state, action: PayloadAction<UploadState>) {
      state.push({ ...action.payload, status: "idle", progress: 0 });
    },
    removeUpload(state, action: PayloadAction<string>) {
      const index = state.findIndex((upload) => upload.id === action.payload);
      if (index !== -1) {
        state.splice(index, 1);
      }
    },
    cleanAllUploads(state) {
      return [];
    },
    cleanFinishedUploads(state) {
      const uploads = state.filter(
        (upload) => upload.status !== "success" && upload.status !== "failed"
      );
      state.splice(0, state.length);
      state.push(...uploads);
    },
    setUploadProgress(state, action: PayloadAction<UploadProgress>) {
      const { id, progress } = action.payload;
      const upload = state.find((upload) => upload.id === id);

      if (upload) {
        upload.progress = progress;
      }
    },
  },
  extraReducers: (builder) => {
    builder.addCase(updateVideo.pending, (state, action) => {
      const upload = state.find((upload) => upload.id === action.meta.arg.id);
      if (upload) {
        upload.status = "loading";
      }
    });

    builder.addCase(updateVideo.fulfilled, (state, action) => {
      const upload = state.find((upload) => upload.id === action.meta.arg.id);
      if (upload) {
        upload.status = "success";
      }
    });

    builder.addCase(updateVideo.rejected, (state, action) => {
      const upload = state.find((upload) => upload.id === action.meta.arg.id);
      if (upload) {
        upload.status = "failed";
      }
    });
  },
});

export const {
  addUpload,
  removeUpload,
  setUploadProgress,
  cleanFinishedUploads,
  cleanAllUploads,
} = uploadsSlice.actions;

// selector
export const selectUploads = (state: RootState) => state.uploadSlice;

export const uploadReducer = uploadsSlice.reducer;
