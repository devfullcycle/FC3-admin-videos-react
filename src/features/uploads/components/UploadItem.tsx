import ArticleIcon from "@mui/icons-material/Article";
import { Box, ListItem, Typography } from "@mui/material";
import { UploadState } from "../UploadSlice";
import { UploadStatus } from "./UploadStatus";

type Props = {
  upload: UploadState;
};

export const UploadItem: React.FC<Props> = ({ upload }) => {
  return (
    <ListItem>
      <Box display={"flex"} justifyContent={"space-between"} width={"100%"}>
        <Box display={"flex"} alignItems={"center"} gap={2}>
          <ArticleIcon />
          <Typography maxWidth={280} noWrap>
            {upload.field}
          </Typography>
        </Box>
        <UploadStatus status={upload.status} progress={upload.progress} />
      </Box>
    </ListItem>
  );
};
