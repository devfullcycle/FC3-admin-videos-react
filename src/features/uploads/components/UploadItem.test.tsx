import { render } from "@testing-library/react";
import { STATUS } from "../UploadSlice";
import { UploadItem } from "./UploadItem";

const Props = {
  upload: {
    id: "test",
    videoId: "test",
    field: "test",
    progress: 0,
    status: STATUS.IDDLE,
    file: new File(["test"], "test.txt", { type: "text/plain" }),
  },
};

describe("UploadItem", () => {
  it("should render UploadItem", () => {
    const { asFragment } = render(<UploadItem upload={Props.upload} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render failed UploadItem", () => {
    const upload = { ...Props.upload, status: STATUS.FAILED };
    const { asFragment } = render(<UploadItem upload={upload} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render uploading UploadItem", () => {
    const upload = { ...Props.upload, status: STATUS.SUCCESS };
    const { asFragment } = render(<UploadItem upload={upload} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
