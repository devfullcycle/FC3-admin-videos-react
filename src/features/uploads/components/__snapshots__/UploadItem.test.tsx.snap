// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadItem should render UploadItem 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
  >
    <div
      class="MuiBox-root css-z7mtfw"
    >
      <div
        class="MuiBox-root css-j0ozid"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArticleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-5 14H7v-2h7v2zm3-4H7v-2h10v2zm0-4H7V7h10v2z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-1y53svw-MuiTypography-root"
        >
          test
        </p>
      </div>
      <span
        aria-valuenow="0"
        class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorSecondary css-1bx0sm6-MuiCircularProgress-root"
        role="progressbar"
        style="width: 20px; height: 20px; transform: rotate(-90deg);"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
            style="stroke-dasharray: 126.920; stroke-dashoffset: 126.920px;"
          />
        </svg>
      </span>
    </div>
  </li>
</DocumentFragment>
`;

exports[`UploadItem should render failed UploadItem 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
  >
    <div
      class="MuiBox-root css-z7mtfw"
    >
      <div
        class="MuiBox-root css-j0ozid"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArticleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-5 14H7v-2h7v2zm3-4H7v-2h10v2zm0-4H7V7h10v2z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-1y53svw-MuiTypography-root"
        >
          test
        </p>
      </div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium css-1wxstni-MuiSvgIcon-root"
        data-testid="ErrorOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
  </li>
</DocumentFragment>
`;

exports[`UploadItem should render uploading UploadItem 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
  >
    <div
      class="MuiBox-root css-z7mtfw"
    >
      <div
        class="MuiBox-root css-j0ozid"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArticleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-5 14H7v-2h7v2zm3-4H7v-2h10v2zm0-4H7V7h10v2z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-1y53svw-MuiTypography-root"
        >
          test
        </p>
      </div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
        data-testid="CheckCircleOutlineIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M16.59 7.58 10 14.17l-3.59-3.58L5 12l5 5 8-8zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
        />
      </svg>
    </div>
  </li>
</DocumentFragment>
`;
